README
*** Overview ***

PostChecker is a distributed monitoring tool for Postfix servers.  The main purpose of the tool is to connect to multiple Postfix servers, check their queue status, and generate a aggregate report based on those numbers.  While PostChecker can be used to monitor a single server, it really starts to make sense when three or more servers need to be monitored.

PostChecker is designed to be a interactive monitoring tool, rather than a trending or capacity planning tool.  However, you can use the "--parseable" option to postchecker_cli.php to generate parseable output you can then feed into another tool for trending purposes, if that's your bag.

Author - Matt Willman - willman.matt@threepaw.com

*** Design ***

PostChecker is split into two primary pieces.  The "client" side code sits on the Postfix servers you want to monitor.  When someone connects to the "client" port, the client side code returns the number of active items in the configured queue directories.

The "server" side consists of three different front ends.

Use:

postchecker_web.php
	-- if you need to check the queues, and get a real-time response
postchecker_cli.php
	-- if you're not really a fan of that whole web thing
	-- or if you want to get parseable output for another tool
postchecker_server.php
	-- to get a continually updating view of your queues

While the server side code can run on the same machine as the client side code, the machine running the server side code accepts the additional burden of processing the statistics returned by the client side code.

*** Install ***

See the included INSTALL document in the primary source tree.

*** Use ***

The INSTALL document details what to put where.  Once everything is installed, the front-ends should produce output in the following manner.

A.  postchecker_web.php should return a dynamic HTML page with the current status of the queues when you invoke it
B.  postchecker_cli.php should return a text output representation of the current status of the queues
C.  postchecker_server.php (when running as a daemon) should update the postchecker.html page (actually the graphs contained in the page) every $update seconds (60 by default)


CHANGELOG
CHANGELOG

Copyright (c) 2005 Matt Willman
ALL RIGHTS RESERVED

Contact: willman.matt@threepaw.com

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc., 59 Temple
Place, Suite 330, Boston, MA 02111-1307 USA

Initial release
07/2005	MBWillman

FORMATTED FOR 80x25

1.3 -	updated for PHP 5

1.2 -	queue_counts.c available for client side
	some minor documentation && directory structure changes

1.0 -	initial release


postchecker_cli.php
#!/usr/bin/php -q

<?php
/*
 * postchecker_cli.php
 *
 * CLI frontend to PostChecker.  Accepts one command line arg to present it's
 * output in machine parseable format.  Otherwise the output is human readable.
 *
 * ./postchecker_cli.php
 *	-- will generate human readable output
 *
 * ./postchecker_cli.php --parseable
 *	-- will generate machine parseable output for use in other tools
 *
 * Author - Matt Willman - willman.matt@threepaw.com
 */

// includes
require_once('/path/to/postchecker_core.php'); 

// defaults
$parseable = false;
$stats = array();
$trash = "";

// parse config
$servers = parse_config($trash);

// check command line args
if (isset($_SERVER['argv'][1]))
{
	if ($_SERVER['argv'][1] == '--parseable')
	{
		$parseable = true;
	}
}

// get the current queue counts
$stats = get_current_counts($stats, $servers);

// human readable gets a nice title
if (!$parseable)
{
	print "PostChecker 1.2\n";

	echo "Queue status as of ->";
	print strftime('%c');
	echo "<-\n\n";
}

// walk through the servers and print out the numbers
foreach ($stats as $server => $trash)
{
	if ($parseable)
	{
		print "$server:";
	}
	else
	{
		print "Server: $server";
	}

	foreach ($stats[$server] as $queue => $trash)
	{
		foreach ($stats[$server][$queue] as $number)
		{
			if ($parseable)
			{
				echo "$queue:$number:";
			}
			else
			{
				echo "  Queue: $queue = $number";
			}
		}
	}

	print "\n";
}// foreach ($stats as $server => $trash)

?>


postchecker_web.php
<?php
/*
 * postchecker_web.php
 *
 * Web front end to PostChecker.  Generates a one-time HTML page that generates graphs depicting the current queues
 *
 * Author - Matt Willman - willman.matt@threepaw.com
 *
 */

// includes
require_once('/path/to/postchecker_core.php'); 

// defaults
$stats = array();
$trash = "";

// parse config file
$servers = parse_config($trash);

// generate unique filenames
$unique = rand();

// get the current queue counts
$stats = get_current_counts($stats, $servers);

// start web page generation
echo "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.01 Transitional//EN\"\n";
echo "   \"http://www.w3.org/TR/html4/loose.dtd\">\n";
echo "<html>\n";
echo "<head>\n";
echo "<title>PostChecker - 1.2</title>\n";
echo "</head>\n";
echo "<body>\n";
echo "<h1>PostChecker - 1.2</h1>\n";
echo "<hr noshade>\n";

echo "Queue status as of ->";
print strftime('%c');
echo "<- <br><br>\n";

// generate the graphs
foreach ($stats as $server => $trash)
{
	echo "Server: $server\n";
	echo "<center>\n";

	foreach ($stats[$server] as $queue => $number)
	{
		if ($queue != 'red_queue')
		{
			// set the filename
			$filename = "$server.$queue.$unique";

			// make the graph for the active queue
			generate_graph($queue, $number, $stats[$server]['red_queue'][0], "$pc_cache/$filename");

			// include it in the web page
			echo "<img src=\"$pc_webroot/$filename.png\"";
			echo " alt=\"$number[0] items in the $queue queue\">\n";
		}
	}

	echo "</center>\n";
	echo "<br>\n";
}

// end web page generation
echo "</body>\n";
echo "</html>\n";

?>


client/queue_counts.c
/*
 * queue_counts.c
 *
 * Alternative client side daemon for PostChecker.  Returns the number of items in the configured spool directories
 * 
 * Author - Matt Willman - willman.matt@threepaw.com
 */


// includes
#include <arpa/inet.h>
#include <dirent.h>
#include <errno.h>
#include <netinet/in.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
#include <syslog.h>

// defines
#define PC_PORT 7678
#define RESTRICTED 1

// functions
int dircount(char *);

// main
int main()
{
	int i, current_count, sock_fd, client_fd, sin_size, ip_matches;

	int yes = 1;

	char spool[] = "/var/spool/postfix/";
	char *dirs[] = {"active", "deferred", "corrupt", "hold"};
	char *restricted[] = {"0.0.0.0"};
	char *current_dir, *current_string;

	struct sockaddr_in my_addr;
	struct sockaddr_in client_addr;

	pid_t pid;

	current_dir = malloc(255);
	current_string = malloc(255);

	// open the connection to syslog
	openlog("queue_counts", LOG_PID, LOG_LOCAL1);

	// security is a good thing
	if (RESTRICTED && (strcmp("0.0.0.0", restricted[0]) == 0))
	{
		syslog(LOG_ERR, "you must configure a server IP address, exiting...\n");
		exit(-1);
	}

	// fork off
	if ((pid = fork()) < 0)
	{
		syslog(LOG_ERR, "unable to fork new child process, exiting...\n");
		exit(-1);
	}
	else if (pid != 0)
	{
		// parent exits
		syslog(LOG_INFO, "starting normal execution\n");
		exit(0);
	}

	// child

	// session leader
	setsid();
	
	// change working directory
	chdir("/");

	// clear umask
	umask(0);

	// open the socket
	if ((sock_fd = socket(AF_INET, SOCK_STREAM, 0)) == -1)
	{
		syslog(LOG_ERR, "unable to open socket %d, exiting...", PC_PORT);
		exit(-1);
	}

	// boilerplate
	my_addr.sin_family = AF_INET;
	my_addr.sin_port = htons(PC_PORT);
	my_addr.sin_addr.s_addr = htonl(INADDR_ANY);
	memset(&(my_addr.sin_zero), '\0', 8);

	// allow socket reuse
	if (setsockopt(sock_fd, SOL_SOCKET, SO_REUSEADDR, &yes,sizeof(int)) == -1)
	{
		syslog(LOG_ERR, "unable to set socket options, exiting...");
		exit(-1);
	}

	// bind to the socket
	if (bind(sock_fd, (struct sockaddr *)&my_addr, sizeof(struct sockaddr)) == -1)
	{
		syslog(LOG_ERR, "unable to bind to socket %d, exiting...", PC_PORT);
		exit(-1);
	}

	// maximum of 5 backlog
	if (listen(sock_fd, 5) == -1)
	{
		syslog(LOG_ERR, "unable to set backlog queue on socket %d, exiting...", PC_PORT);
		exit(-1);
	}

	// sit here and wait
	while(1)
	{
		sin_size = sizeof(struct sockaddr);
		if ((client_fd = accept(sock_fd, (struct sockaddr *)&client_addr, &sin_size)) == -1)
		{
			// couldn't accept for some reason
			syslog(LOG_WARNING, "unable to accept connection from %s on socket %d, resuming operation", inet_ntoa(client_addr.sin_addr), PC_PORT);
			continue;
		}
		else
		{
			// fake out the check below if we're not in RESTRICTED mode
			ip_matches = 1;

			if (RESTRICTED)
			{
				ip_matches = 0;

				// allow for multiple connecting IP addresses
				for (i = 0; i < (sizeof(restricted)/sizeof(restricted[0])); i++)
				{
					if (strcmp(restricted[i], inet_ntoa(client_addr.sin_addr)) == 0)
					{
						ip_matches = 1;
					}
				}

				if (!ip_matches)
				{
					// if IP's don't match, close the socket
					syslog(LOG_ERR, "rejected connection from %s, resuming operation", inet_ntoa(client_addr.sin_addr)); 
					close(client_fd);
				}
				else
				{
					syslog(LOG_INFO, "accepted connection from %s", inet_ntoa(client_addr.sin_addr));
				}
			}

			// if we reject the connection, don't bother checking the queues
			if (ip_matches)
			{
				// you should be able to change the spool directories
				// without changing the loop parameters
				for (i = 0; i < (sizeof(dirs)/sizeof(dirs[0])); i++)
				{
					// construct the current queue dir
					current_dir = strcpy(current_dir, spool);
					current_dir = strcat(current_dir, dirs[i]);

					// get the count
					current_count = dircount(current_dir);

					// format the output string
					sprintf(current_string, "%s:%d\n", dirs[i], current_count);

					// send it back to the client
					send(client_fd, current_string, strlen(current_string), 0);
				}

				// close the connection
				close(client_fd);
			}
		}
	}// while(1)
}// int main()


/*
 * int dircount(char *)
 *
 * accept a directory name, and return the number of files within it
 */

int dircount(char *current_dir)
{
	int count = 0;

	struct dirent *dp;
	DIR *dfd;

	// open the directory
	if ((dfd = opendir(current_dir)) == NULL)
	{
		syslog(LOG_ERR, "unable to open spool directory %s, skipping", current_dir);
		return(-1);
	}

	// iterate through the directory and count up the files
	while ((dp = readdir(dfd)) != NULL)
	{
		if (strcmp(dp->d_name, ".") == 0
			|| strcmp(dp->d_name, "..") == 0)
		{
			continue;
		}

		count++;
	}

	// be a good citizen
	closedir(dfd);

	return(count);
}// int dircount(char *current_dir)


postchecker_server.php
#!/usr/bin/php -q

<?php
/*
 * postchecker_server.php
 *
 * Server side HTML generator for PostChecker.  Checks the configured machines * every $update seconds, and regenerates the graphs.  Keeps the last five * results returned in the graphs by default.
 *
 * Author - Matt Willman - willman.matt@threepaw.com
 */

// includes
require_once('/path/to/postchecker_core.php'); 

// defaults
$stats = array();

// parse config
$servers = parse_config(&$update);

// get current queue counts
$stats = get_current_counts($stats, $servers);

// open the html file
$file = fopen("$pc_cache/postchecker.html", 'w') or trigger_error("cannot open file for writing", E_USER_ERROR);

// start web page generation
fwrite($file, "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.01 Transitional//EN\"\n");
fwrite($file, "   \"http://www.w3.org/TR/html4/loose.dtd\">\n");
fwrite($file, "<html>\n");
fwrite($file, "<head>\n");
fwrite($file, "<meta http-equiv=\"refresh\" content=\"$update\">\n");
fwrite($file, "<title>PostChecker - 1.2</title>\n");
fwrite($file, "</head>\n");
fwrite($file, "<body>\n");
fwrite($file, "<h1>PostChecker - 1.2</h1>\n");
fwrite($file, "This page is updated every $update seconds<br>\n");
fwrite($file, "<hr noshade>\n");

// generate the graphs
foreach ($stats as $server => $trash)
{
	fwrite($file, "Server: $server\n");
	fwrite($file, "<center>\n");

	foreach ($stats[$server] as $queue => $number)
	{
		if ($queue != 'red_queue')
		{
			// set the filename
			$filename = "$server.$queue";

			// make the graph for the active queue
			generate_graph($server, $queue, $number, $stats[$server]['red_queue'][0], "$pc_cache/$filename");

			// make sure the webserver can read it
			chmod("$pc_cache/$filename.png", 0755);

			// include it in the web page
			fwrite($file, "<img src=\"$pc_webroot/$filename.png\"");

			// we're not updating the web page (just the graphs)
			// so we don't always have the current number
			fwrite($file, " alt=\"number of items in the $queue queue\">\n");
		}
	}

	fwrite($file, "</center>\n");
	fwrite($file, "<br>\n");
}

// end web page generation
fwrite($file, "</body>\n");
fwrite($file, "</html>\n");

// flush and close
fflush($file);
fclose($file);

// make sure the webserver can read it
chmod("$pc_cache/postchecker.html", 0755);

// re-generate the graphs every $update seconds
while (true)
{
	sleep($update);

	// update the counts
	$stats = get_current_counts($stats, $servers);

	foreach ($stats as $server => $trash)
	{
		foreach ($stats[$server] as $queue => $number)
		{
			if ($queue != 'red_queue')
			{
				// set the filename
				$filename = "$server.$queue";

				// make the graph for the active queue
				generate_graph($server, $queue, $number, $stats[$server]['red_queue'][0], "$pc_cache/$filename");

				// make sure the webserver can read it
				chmod("$pc_cache/$filename.png", 0755);
			}
		}
	}
}// while(true)

?>


postchecker_core.php
<?php
/*
 * postchecker_core.php
 *
 * Contains all the core functionality used by the frontends
 *
 * Author - Matt Willman - willman.matt@threepaw.com
 *
 * Initial release
 * 07/2005	MBWillman
 *
 * Updated for PHP 5 - you must use a version of JPgraph that is PHP 5 compliant
 * 05/2008	MBWillman
 *
 * Remove unused variable in generate_graphs()
 * 07/2008	MBWillman
 */


// includes
include ("/path/to/jpgraph.php");
include ("/path/to/jpgraph_bar.php");

// defaults
$pc_cache = "/path/to/writeable/apache/temp/directory";
$pc_webroot = "/truncated/web/path/to/$pc_cache";
$debug = false;

// set local error handler
set_error_handler('error_hand');

// set local time zone
date_default_timezone_set("America/Los_Angeles");

/*
 * function parse_config()
 *
 * parse the configuration file
 */

function parse_config($update)
{
	// default to 60 if not configured
	$default_update = 60;

	// default to 2500 if not configured
	$red_queue = 25;

	// location of the config file
	$config_file = "/path/to/postchecker.cfg";

	// read in the config file
	if (fopen($config_file, 'r'))
	{
		$config = parse_ini_file($config_file);

		// set $update if !isset
		if (!(isset($config['update'])))
		{
			$update = $default_update;
		}
		else
		{
			$update = $config['update'];
		}

		// set $red_queue if !isset
		if (!(isset($config['red_queue'])))
		{
			$config['red_queue'] = $red_queue;
		}

		// that's not good
		if (count($config) == 0)
		{
			trigger_error("no monitored servers configured", E_USER_ERROR);
		}

		// get the servers
		reset($config);

		while (list($key, $red_queue) = each($config))
		{
			if (!(preg_match('/(update|red_queue)/', $key)))
			{
				if ($config[$key] == "")
				{
					$config[$key] = $config['red_queue'];
				}

				// populate the server array
				$servers[$key] = $config[$key];
			}
		}

		if (count($servers) == 0)
		{
			trigger_error("no monitored servers configured", E_USER_ERROR);
		}
	}
	else
	{
		trigger_error("cannot open configuration file", E_USER_ERROR);
	}

	return $servers;
}// parse_config($update)


/*
 * get_current_counts()
 *
 * get the current queue counts from all the configured servers
 */

function get_current_counts($stats, $servers)
{
	// connect to all configured machines
	while (list($key, $red_queue) = each($servers))
	{
		$socket = fsockopen($key, 7678, $errno, $errstr, 10);

		if (!$socket)
		{
			// whoopsie
			trigger_error("unable to open socket", E_USER_ERROR);
		}
		else
		{
			while (!feof($socket))
			{
				$current = rtrim(fgets($socket));

				if ($current)
				{
					// parse the input
					list($queue, $number) = explode(":", $current);

					// update the queue numbers
					if (isset($stats[$key][$queue]))
					{
						// make sure we shift the array
						if (count($stats[$key][$queue]) == 5)
						{
							$trash = array_shift($stats[$key][$queue]);
						}
							
						// and now add it in
						$stats[$key][$queue] = array_merge($stats[$key][$queue], array($number));
					}
					else
					{	
						// first time ever through the loop
						$stats[$key][$queue] = array($number);
					}
				}
			}

			fclose($socket);

			// add the red_queue value
			$stats[$key]['red_queue'] = array($red_queue);
		}
	}// while (list($key, $red_queue) = each($servers))

	return $stats;
}// get_current_counts()


/*
 * generate_graph($l_queue, $l_number, $l_red_queue, $filename)
 *
 * generate the graph from input
 */

function generate_graph($l_queue, $l_number, $l_red_queue, $filename)
{
	// graph setup
	$current_graph = new Graph(200,200);
	$current_graph->SetScale("textint");
	$current_graph->SetShadow();
	$current_graph->yaxis->scale->SetGrace(10);
	$current_graph->SetMargin(50,20,0,40);

	// for every number in array($l_number), create a barplot
	foreach ($l_number as $current_number)
	{
		$current_array = array($current_number);

		$current_plot = new BarPlot($current_array);
		$current_plot->value->Show();
		$current_plot->value->SetFormat('%d');

		// prettify the formatting
		if ($current_number != 0)
		{
			$current_plot->SetValuePos('center');
		}
		
		// are we green, yellow, or red?
		if ($current_number >= $l_red_queue)
		{
			$current_plot->SetFillColor('red');
		}
		else
		{
			if ($current_number >= ($l_red_queue * .75))
			{
				$current_plot->SetFillColor('yellow');
			}
			else
			{
				$current_plot->SetFillColor('green');
			}
		}

		// first time through the loop?
		if (isset($current_group_array))
		{
			$current_group_array = array_merge($current_group_array, array($current_plot));
		}
		else
		{
			$current_group_array = array($current_plot);
		}
	}

	// add up all the barplots
	$current_group_plot = new GroupBarPlot($current_group_array);
	$current_group_plot->SetWidth(0.8);
	$current_graph->Add($current_group_plot);

	$current_graph->title->Set($l_queue);
	$current_graph->yaxis->title->Set("Items");
	$current_graph->yaxis->title->SetMargin(10);

	// assumes your default output is png
	$current_graph->Stroke("$filename.png");

}// generate_graph($l_queue, $l_number, $l_red_queue, $filename)


/*
 * function error_hand($errno, $errstr, $errfile, $errline)
 *
 * standard error handling function
 */

function error_hand($errno, $errstr, $errfile, $errline)
{
	global $debug;

	if ($debug)
	{
		echo "errno: $errno<br>";
		echo "errstr: $errstr<br>";
		echo "errfile: $errfile<br>";
		echo "errline: $errline<br>";
		
		exit(1);
	}

	echo "FATAL ERROR: $errstr\n";
	exit(1);

}// function error_hand($errno, $errstr, $errfile, $errline)

?>
